File: src/lib/research/extractive.ts
import type { ScrapedSource } from "./types";
import { tokenize } from "./text";

export function extractiveReport(topic: string, sources: ScrapedSource[]): string {
  const bundles = sources.filter((s) => s.content.trim()).map(sourceSentences);
  const all: Array<[string, string]> = [];
  for (const bundle of bundles) {
    for (const sentence of bundle.sentences) all.push([sentence, bundle.title]);
  }
  const selected = selectDiverse(topic, all, 18);
  const summary = joinUnique(selected.slice(0, 4));
  const findings = selected.length > 6 ? selected.slice(1, 7) : selected.slice(0, 5);

  const parts: string[] = [
    `# Research Report: ${topic}`,
    "",
    "## Executive Summary",
    "",
    summary || "The selected sources did not contain enough overlapping, usable prose to summarise.",
    "",
    "## Key Findings",
    "",
  ];

  findings.forEach(([sentence, title], i) => {
    parts.push(`### Finding ${i + 1}`, "", `${sentence} _(Source: ${title})_`, "");
  });

  parts.push(
    "## Detailed Analysis",
    "",
    "The notes below stay close to the extracted source text. Repeated claims across pages were dropped so each point appears once.",
    "",
  );

  const used = new Set(findings.map(([s]) => s));
  for (const bundle of bundles) {
    const leftover = dedupeList(bundle.sentences.filter((s) => !used.has(s))).slice(0, 6);
    if (!leftover.length) continue;
    parts.push(`### ${bundle.title}`, "", leftover.join(" "), "");
    leftover.forEach((s) => used.add(s));
  }

  const conclusionPool = selected.length >= 3 ? selected.slice(-3) : selected;
  const conclusion =
    joinUnique(conclusionPool) ||
    `Across the collected sources, the material on “${topic}” is limited to the pages listed below.`;
  parts.push("## Conclusion", "", conclusion, "");
  return `${parts.join("\n").trim()}\n`;
}

function sourceSentences(source: ScrapedSource): { title: string; sentences: string[] } {
  const text = source.content.replace(/^\(Full page could not be extracted:[^)]+\)\s*/, "").replace(/\n/g, " ");
  const raw = text.split(/(?<=[.!?])\s+(?=[A-Z“"])/);
  const sentences = raw
    .map((piece) => piece.replace(/\s+/g, " ").trim())
    .filter((piece) => piece.length >= 60 && piece.length <= 320 && !piece.startsWith("http"));
  return { title: source.title, sentences: sentences.slice(0, 40) };
}

function selectDiverse(topic: string, items: Array<[string, string]>, k: number): Array<[string, string]> {
  const terms = tokenize(topic);
  const scored = items
    .map(([sentence, title]) => {
      const tokens = tokenize(sentence);
      const overlap = tokens.size ? [...terms].filter((t) => tokens.has(t)).length / Math.max(terms.size, 1) : 0;
      const density = Math.min(sentence.length / 180, 1);
      return { score: 0.7 * overlap + 0.3 * density, sentence, title };
    })
    .sort((a, b) => b.score - a.score);

  const picked: Array<[string, string]> = [];
  for (const item of scored) {
    if (picked.some(([existing]) => tooSimilar(item.sentence, existing))) continue;
    picked.push([item.sentence, item.title]);
    if (picked.length >= k) break;
  }
  return picked;
}

function tooSimilar(a: string, b: string, threshold = 0.55): boolean {
  const ta = tokenize(a);
  const tb = tokenize(b);
  if (!ta.size || !tb.size) return false;
  let inter = 0;
  for (const t of ta) if (tb.has(t)) inter += 1;
  return inter / (ta.size + tb.size - inter) >= threshold;
}

function dedupeList(sentences: string[]): string[] {
  const kept: string[] = [];
  for (const sentence of sentences) {
    if (kept.some((existing) => tooSimilar(sentence, existing))) continue;
    kept.push(sentence);
  }
  return kept;
}

function joinUnique(items: Array<[string, string]>): string {
  const kept: Array<[string, string]> = [];
  for (const item of items) {
    if (kept.some(([existing]) => tooSimilar(item[0], existing))) continue;
    kept.push(item);
  }
  return kept.map(([s]) => s).join(" ");
}

File: GA_8_AutonomousResearchAgent_BYTE/src/extractive.py
"""Extractive fallback: a structured report from source sentences only.

Used when the LLM API is missing or rejects the request. Nothing is invented —
sentences come from the scraped pages, duplicates are dropped by token overlap.
"""

from __future__ import annotations

import re

from .scraper import ScrapedSource
from .utils import tokenize

_SENT_SPLIT = re.compile(r"(?<=[.!?])\s+(?=[A-Z“\"])")
_JUNK = re.compile(
    r"(article sidebar|main article content|cookie|subscribe|sign in|log in|"
    r"accept all|related articles|share this|advertisement)",
    re.I,
)


def extractive_report(topic: str, sources: list[ScrapedSource]) -> str:
    bundles = [_source_sentences(source) for source in sources if source.content.strip()]
    all_sents: list[tuple[str, str]] = [(s, b["title"]) for b in bundles for s in b["sentences"]]
    selected = _select_diverse(topic, all_sents, k=18)

    summary = _join_unique(selected[:4])
    findings = selected[1:7] if len(selected) > 6 else selected[:5]
    if not findings:
        findings = selected[:3]

    parts = [
        f"# Research Report: {topic}",
        "",
        "## Executive Summary",
        "",
        summary or "The selected sources did not contain enough overlapping, usable prose to summarise.",
        "",
        "## Key Findings",
        "",
    ]
    for i, (sentence, title) in enumerate(findings, start=1):
        parts.append(f"### Finding {i}")
        parts.append("")
        parts.append(f"{sentence} _(Source: {title})_")
        parts.append("")

    parts.append("## Detailed Analysis")
    parts.append("")
    parts.append(
        "The notes below stay close to the extracted source text. "
        "Repeated claims across pages were dropped so each point appears once."
    )
    parts.append("")

    used = {s for s, _ in findings}
    for bundle in bundles:
        leftover = [s for s in bundle["sentences"] if s not in used]
        leftover = _dedupe_list(leftover)[:6]
        if not leftover:
            continue
        parts.append(f"### {bundle['title']}")
        parts.append("")
        parts.append(" ".join(leftover))
        parts.append("")
        used.update(leftover)

    conclusion_pool = selected[-3:] if len(selected) >= 3 else selected
    conclusion = _join_unique(conclusion_pool)
    parts.append("## Conclusion")
    parts.append("")
    parts.append(
        conclusion
        or f"Across the collected sources, the material on “{topic}” is limited to the pages listed below."
    )
    parts.append("")
    return "\n".join(parts).strip() + "\n"


def _clean_sentence(piece: str) -> str:
    piece = re.sub(r"\s+", " ", piece).strip(" \t.:;)(")
    if piece.startswith(")"):
        piece = piece.lstrip(") ").strip()
    if not piece.endswith((".", "!", "?")):
        piece = piece + "."
    return piece[0].upper() + piece[1:] if piece else piece


def _is_junk(piece: str) -> bool:
    if len(piece) < 60 or len(piece) > 320:
        return True
    if piece.lower().startswith("http"):
        return True
    if _JUNK.search(piece):
        return True
    if len(piece.split()) < 10:
        return True
    return False


def _source_sentences(source: ScrapedSource) -> dict:
    text = source.content
    text = re.sub(r"^\(Full page could not be extracted:[^)]+\)\s*", "", text)
    raw = _SENT_SPLIT.split(text.replace("\n", " "))
    sentences = []
    for piece in raw:
        piece = _clean_sentence(piece)
        if not _is_junk(piece):
            sentences.append(piece)
    return {"title": source.title, "url": source.url, "sentences": sentences[:40]}


def _select_diverse(topic: str, items: list[tuple[str, str]], k: int) -> list[tuple[str, str]]:
    terms = tokenize(topic)
    scored: list[tuple[float, str, str]] = []
    for sentence, title in items:
        tokens = tokenize(sentence)
        if not tokens:
            continue
        overlap = len(terms & tokens) / max(len(terms), 1)
        density = min(len(sentence) / 180, 1.0)
        scored.append((0.7 * overlap + 0.3 * density, sentence, title))
    scored.sort(reverse=True)

    picked: list[tuple[str, str]] = []
    for _, sentence, title in scored:
        if any(_too_similar(sentence, existing) for existing, _ in picked):
            continue
        picked.append((sentence, title))
        if len(picked) >= k:
            break
    return picked


def _too_similar(a: str, b: str, threshold: float = 0.55) -> bool:
    ta, tb = tokenize(a), tokenize(b)
    if not ta or not tb:
        return False
    return len(ta & tb) / len(ta | tb) >= threshold


def _dedupe_list(sentences: list[str]) -> list[str]:
    kept: list[str] = []
    for sentence in sentences:
        if any(_too_similar(sentence, existing) for existing in kept):
            continue
        kept.append(sentence)
    return kept


def _join_unique(items: list[tuple[str, str]]) -> str:
    return " ".join(sentence for sentence, _ in _dedupe_list_pairs(items))


def _dedupe_list_pairs(items: list[tuple[str, str]]) -> list[tuple[str, str]]:
    kept: list[tuple[str, str]] = []
    for sentence, title in items:
        if any(_too_similar(sentence, existing) for existing, _ in kept):
            continue
        kept.append((sentence, title))
    return kept
